model/pachinko_gibbs_lda.py

```diff
@@ -0,0 +1,266 @@
+import pickle
+from sklearn.model_selection import train_test_split
+import time
+from typing import List
+from scipy import sparse
+import numpy as np
+from tqdm import tqdm
+import random
+
+import utility
+from gibbs_utility import perplexity, get_coherence, mean_topic_diff, get_topics, decrease_count, increase_count, \
+    cat_perplexity
+from preprocess.preprocessing import prepro_file_load
+import math
+
+
+def random_initialize(documents):
+    """
+    Randomly initialisation of the word topics
+    :param documents: a list of documents with their word ids
+    :return: word_topic_assignment,
+    """
+    print("Random Initilization")
+    np.random.seed()
+    middle_layers = [[] for _ in range(N)]
+    topic_to_word = np.zeros([layer_lengths[len(layer_lengths)-1], M])
+    word_topic_assignment = []
+    for d, doc in tqdm(documents):
+        mid_doc_layers = [np.zeros((layer_lengths[x], layer_lengths[x+1]), dtype=np.intc) for x in range(len(layer_lengths)-1)]
+        tax_ids = doc2tax[d]
+        tax = tax2topic_id(tax_ids)
+        currdoc = []
+        for w in doc:
+            # randomly choose a taxonomy chain if there are multiple
+            if len(tax) > 1:
+                word_tax = [x for x in tax[random.randrange(len(tax))]]
+            elif len(tax) == 1:
+                word_tax = [x for x in tax[0]]
+            else:
+                word_tax = []
+
+            while len(word_tax) < len(layer_lengths):
+                word_tax.append(np.random.randint(layer_lengths[len(word_tax)]))
+            z = tuple(word_tax)
+            # apply various counts
+            for i, matrix in enumerate(mid_doc_layers):
+                matrix[z[i], z[i+1]] += 1
+            currdoc.append(z)
+            topic_to_word[z[len(z)-1], w] += 1
+        word_topic_assignment.append(currdoc)
+        middle_layers[d] = mid_doc_layers
+    return word_topic_assignment, middle_layers, topic_to_word
+
+
+def decrease_counts(topics, middle_layers, middle_sums, topic_to_word, topic_to_word_sums, word, d):
+    for i, x in enumerate(middle_layers[d]):
+        x[topics[i], topics[i+1]] -= 1
+    for i, x in enumerate(middle_sums):
+        x[topics[i]] -= 1
+    topic_to_word[topics[len(topics)-1], word] -= 1
+    topic_to_word_sums[topics[len(topics)-1]] -= 1
+
+
+def increase_counts(topics, middle_layers, middle_sums, topic_to_word, topic_to_word_sums, word, d):
+    for i, x in enumerate(middle_layers[d]):
+        x[topics[i], topics[i + 1]] += 1
+    for i, x in enumerate(middle_sums):
+        x[topics[i]] += 1
+    topic_to_word[topics[len(topics) - 1], word] += 1
+    topic_to_word_sums[topics[len(topics) - 1]] += 1
+
+
+def gibbs_sampling(documents: List[np.ndarray],
+                   wta,
```

Comment:
Vil nok hellere have det skrevet ud, da det alligevel kun bruges 1 sted i koden

Reply:
altså i main? så bliver main nok meget lang, men det kan vi godt?

Reply:
```suggestion
                   word_topic_assignment,
```
Jeg tænkte sådan her, og så ændre den på linje 113. Linje 174 bruger nemlig også lige nu word_topic_assignment fra main.

Reply:
right forstår hvad du mener nu.